### PR TITLE
fix: add "next_scheduled_time"

### LINKED ISSUE #17399: Add  Next scheduled time to tag-retention schedule
1. When we set up a cron job, the UI should display the next run time, so that the user can know the next execution time in advance and prepare for it
image

### PR DESCRIPTION
**Pull request by zyyw**
Signed-off-by: Shengwen Yu <yshengwen@vmware.com>

Thank you for contributing to Harbor!

# Comprehensive Summary of your change

# Issue being fixed
Fixes #17399

Please indicate you've done the following:
- [ ] Well Written Title and Summary of the PR
- [ ] Label the PR as needed. "release-note/ignore-for-release, release-note/new-feature, release-note/update, release-note/enhancement, release-note/community, release-note/breaking-change, release-note/docs, release-note/infra, release-note/deprecation"
- [ ] Accepted the DCO. Commits without the DCO will delay acceptance.
- [ ] Made sure tests are passing and test coverage is added if needed.
- [ ] Considered the docs impact and opened a new docs issue or PR with docs changes if needed in [website repository](https://github.com/goharbor/website).

### CODE REVIEW

File: src/pkg/retention/manager.go [hunk 1]
```diff
@@ -60,6 +63,7 @@ func (d *DefaultManager) CreatePolicy(ctx context.Context, p *policy.Metadata) (
 
 // UpdatePolicy Update Policy
 func (d *DefaultManager) UpdatePolicy(ctx context.Context, p *policy.Metadata) error {
+	_ = addNextScheduledTime(p)
```
Comment (wy65701436) [inline src/pkg/retention/manager.go:66]: why skip this err?

File: src/pkg/retention/manager.go [hunk 2]
```diff
@@ -99,3 +103,13 @@ func (d *DefaultManager) GetPolicy(ctx context.Context, id int64) (*policy.Metad
 func NewManager() Manager {
 	return &DefaultManager{}
 }
+
+func addNextScheduledTime(p *policy.Metadata) error {
```
Comment (wy65701436) [inline src/pkg/retention/manager.go:107]: no error will be raised?

File: src/pkg/retention/manager.go [hunk 3]
```diff
@@ -99,3 +103,13 @@ func (d *DefaultManager) GetPolicy(ctx context.Context, id int64) (*policy.Metad
...
+	if p.Trigger.Kind == policy.TriggerKindSchedule {
```
Comment (wy65701436) [inline src/pkg/retention/manager.go:108]:
```
if p.Trigger.Kind != policy.TriggerKindSchedule {
     return
}
 ....
```

File: src/pkg/retention/manager.go [hunk 4]
```diff
@@ -99,3 +103,13 @@ func (d *DefaultManager) GetPolicy(ctx context.Context, id int64) (*policy.Metad
...
+		cron, ok := p.Trigger.Settings[policy.TriggerSettingsCron]
+		if ok && len(cron.(string)) > 0 {
+			p.Trigger.Settings[policy.TriggerSettingNextScheduledTime] = strfmt.DateTime(utils.NextSchedule(p.Trigger.Settings[policy.TriggerSettingsCron].(string), time.Now()))
```
Comment (wy65701436) [inline src/pkg/retention/manager.go:111]: `utils.NextSchedule(p.Trigger.Settings[policy.TriggerSettingsCron] => cron?`

Comment (zyyw) [reply src/pkg/retention/manager.go:111]: Yes, verified on hourly, daily, weekly, and custom cron setting.

**Review by stonezdj — APPROVED** [review]

**Review by chlins — APPROVED** [review]
lgtm

**Pull request merged**